swev-id: sphinx-doc__sphinx-7889 Fix autodoc mock TypeError with generics in _make_subclass

casey-brooks opened the pull request:
## Summary
- add origin/type coercion helpers so `_MockObject.__getitem__` accepts generics
- sanitize generated subclass names and use display names for mocked generics
- add unit and integration coverage for Thing[T], Thing[List[T]], Thing[Mapping[K, V]], and Thing[K, V]

## Testing
- `pytest tests/ext/autodoc/test_mock_generics.py -q`
- `pytest tests/test_ext_autodoc_mock_generics.py -q`
- `pytest tests/test_ext_autodoc_mock.py -q`
- `flake8 sphinx/ext/autodoc/mock.py tests/ext/autodoc/test_mock_generics.py tests/test_ext_autodoc_mock_generics.py tests/roots/test-autodoc-mock-generics/conf.py tests/roots/test-autodoc-mock-generics/target.py`

## Reproduction
1. Configure `autodoc_mock_imports = ['mylib']` on branch `sphinx-doc__sphinx-7889`.
2. Autodoc a module containing `class Foo(Thing[T])` or similar where `Thing` originates from the mocked package.
3. The build fails with `TypeError: type() argument 1 must be str, not typing.List[~T]` thrown by `_make_subclass` when Python passes the generic alias to `type()`.

Fixes #34.

casey-brooks committed:
fix(autodoc): handle mocked generics

support sanitized mock generics display names

Refs #34

casey-brooks commented:
Local verification:
- `pytest tests/ext/autodoc/test_mock_generics.py -q` → 4 passed
- `pytest tests/test_ext_autodoc_mock_generics.py -q` → 1 passed
- `pytest tests/test_ext_autodoc_mock.py -q` → 6 passed
- `flake8 sphinx/ext/autodoc/mock.py tests/ext/autodoc/test_mock_generics.py tests/test_ext_autodoc_mock_generics.py tests/roots/test-autodoc-mock-generics/conf.py tests/roots/test-autodoc-mock-generics/target.py` → no issues

noa-lucent commented on sphinx/ext/autodoc/mock.py:
```diff
@@ -68,12 +74,98 @@ def __repr__(self) -> str:
+        else:
+            return base
+
+    if hasattr(value, '__module__') and hasattr(value, '__qualname__'):
```
[major] When the subscription key is another mocked object (e.g. `Thing[OtherThing[T]]`), we will hit this branch. Instances of `_MockObject` set `__qualname__` to an empty string and rely on `__display_name__` for their human-readable name. Returning `f"{module}.{qualname}"` produces values like `"mylib.OtherThing."`, so the resulting display for nested mocked generics becomes `Thing[mylib.OtherThing.]`. That regresses the representation we are trying to fix. Please special-case mocked objects before this block (e.g. check for `getattr(value, '__display_name__', None)`) and use the stored display name when present.

noa-lucent commented on tests/ext/autodoc/test_mock_generics.py:
```diff
@@ -0,0 +1,55 @@
+
+    specialized = thing[K, V]
+
+    assert repr(specialized) == 'mylib.Thing[K, V]'
```
[major] Please add coverage for nested mocked generics (e.g. `thing[thing[T]]`). The current implementation mis-renders that case because  relies on  for mocked objects, and without a test it is easy to regress.

noa-lucent replied:
[major] Please add a test for nested mocked generics (e.g. `thing[thing[T]]`). That scenario currently renders incorrectly because `_stringify` does not consume the stored `__display_name__`. A regression test will keep the fix in place.

noa-lucent reviewed (changes requested):
The new coercion path is close, but I noticed that nested mocked generics still render incorrectly and there is no test covering that scenario. Please see inline comments for details.

casey-brooks committed:
fix(autodoc): preserve mocked display names

use __display_name__ when stringifying mocked generics and cover nested subscriptions

Refs #34

casey-brooks commented:
Addressed feedback:
- `_stringify` now returns a mocked object's `__display_name__` before falling back to module/qualname, fixing nested mocked generics.
- Added `thing[thing[T]]` regression coverage in `tests/ext/autodoc/test_mock_generics.py`.

Verification:
- `pytest tests/ext/autodoc/test_mock_generics.py -q` → 5 passed
- `pytest tests/test_ext_autodoc_mock_generics.py -q` → 1 passed
- `pytest tests/test_ext_autodoc_mock.py -q` → 6 passed
- `flake8 sphinx/ext/autodoc/mock.py tests/ext/autodoc/test_mock_generics.py tests/test_ext_autodoc_mock_generics.py tests/roots/test-autodoc-mock-generics/conf.py tests/roots/test-autodoc-mock-generics/target.py` → no issues

noa-lucent reviewed (approved):
Thanks for addressing the nested mocked generic path—using  in  fixes the regression, and the new regression test protects it going forward. Looks good.
